File: src/0208-Implement-Trie-(Prefix Tree)/0208.py
class Node:
    def __init__(self, isWord=False):
        self.isWord = isWord
        self.next = dict()
        
class Trie:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = Node()

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        cur = self.root
        for c in word:
            if c not in cur.next:
                cur.next[c] = Node()
            cur = cur.next[c]
            
        if not cur.isWord:
            cur.isWord = True
        
    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        cur = self._search(word)
        return cur != None and cur.isWord
        
    def _search(self, word):
        cur = self.root
        for c in word:
            if c not in cur.next:
                return None
            cur = cur.next[c]
            
        return cur
        
    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        return self._search(prefix) != None 

File: src/0208-Implement-Trie-(Prefix-Tree)/0208.go
type Trie struct {
    next map[rune]*Trie
    isWord bool
}


/** Initialize your data structure here. */
func Constructor() Trie {
    return Trie{next: make(map[rune]*Trie)}
}


/** Inserts a word into the trie. */
func (this *Trie) Insert(word string)  {
    cur := this
    for _, c := range word {
        if cur.next[c] == nil {
            cur.next[c] = &Trie{next: make(map[rune]*Trie)}
        }
        cur = cur.next[c]
    }
    if !cur.isWord {
        cur.isWord = true
    }
}


func (this *Trie) search(word string) *Trie {
    cur := this
    for _, c := range word {
        if cur.next[c] == nil {
            return nil
        }
        cur = cur.next[c]
    }
    return cur
}

/** Returns if the word is in the trie. */
func (this *Trie) Search(word string) bool {
    cur := this.search(word)
    if cur != nil && cur.isWord {
        return true
    }
    return false
}


/** Returns if there is any word in the trie that starts with the given prefix. */
func (this *Trie) StartsWith(prefix string) bool {
    if this.search(prefix) != nil {
        return true
    }
    return false
}